### ISSUE TITLE
ReFL implement details

### ISSUE THREAD

**Issue by hkunzhe** [THUDM/ImageReward#57]
As mentioned in #24 and #34, the current [ReFL code](https://github.com/THUDM/ImageReward/blob/main/ImageReward/ReFL.py) only the ReFL loss is implemented and the pre-training loss is not included. In addition, the two losses are optimized alternately.
![](https://user-images.githubusercontent.com/52482160/252149464-a3e138c9-8347-4f31-80d1-ce3cd5fd9d6b.png)

I want to add pre-training data myself. If we don't use the gradient accumulation, the pseudo code would be like this:
```python
# Given optimizer and lr_scheduler with unet.
# Compute Pre-training Loss `train_loss` with unet and update unet.
train_loss.backward()
optimizer.step()
lr_scheduler.step()  # is it necessary?
optimizer.zero_grad()

# Compute ReFL Loss `refl_loss` with unet and update unet.
refl_loss.backward()
optimizer.step()
lr_scheduler.step()
optimizer.zero_grad()
```

However, I'm confused about how to add `accelerator.accumulate(unet)` for gradient accumulation after reading [this post](https://discuss.pytorch.org/t/multiple-gradient-updates-with-two-separate-losses-and-two-classifiers-sharing-the-same-encoder/170207/5). And I also raised the issue huggingface/accelerate#1870 and [discussion](https://discuss.huggingface.co/t/question-how-to-optimize-two-loss-alternately-with-gradient-accumulation/51521) in the huggingface accelerate github repo and forum. But I don't seem to get a clear answer. Can you give me some pseudo codes or hints? Thank you very much! @xujz18 @tongyx361

**Comment by hkunzhe** [comment]
May be duplicated with https://github.com/THUDM/ImageReward/issues/34#issuecomment-1687365733. I was afraid it wouldn't be seen in a closed issue, so I raised this new issue.

**Comment by xujz18** [comment]
Your understanding is correct and I appreciate the discussion.
